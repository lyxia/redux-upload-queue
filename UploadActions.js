//@flow

import {createAction} from 'redux-actions'

import UploadEventEmitter from './UploadEventEmitter'

import {config} from './UploadComponent'

export const pushUploadItem = createAction('Upload/pushItem')
export const deleteUploadItem = createAction('Upload/deleteItem')
export const cleanUpload = createAction('Upload/cleanUpload')
export const startuploadItem = createAction('Upload/startUploadItem')
export const uploadItemFailed = createAction('Upload/failedUploadItem')
export const uploadItemSuccess = createAction('Upload/successUploadItem')

export const startUpload = createAction('Upload/startUpload')
export const uploadComplete = createAction('Upload/uploadComplete')

export const uploadCancle = createAction('Upload/uploadCancle')
export const registerUpload = createAction('Upload/registerUpload')
export const destroyUpload = createAction('Upload/destroyUpload')

export function upload(name, config) {
    return (dispatch, getState) => {
        const state = getState()
        if(state.upload.get(name)) {
            if(!state.upload.get(name).uploading) {
                //开始上传
                dispatch(startUpload({upload: name}))
                return uploadNextItem(name, dispatch, getState, config)
            }
        }
        return Promise.resolve()
    }
}

function uploadItem(name, item, dispatch, getState, config) {
    dispatch(startuploadItem({upload: name}))
    return config.uploadApi(item.filepath, item.name)
        .then((model) => {
            if(config.isSuccess(model)) {
                //fileName
                dispatch(uploadItemSuccess({upload: name, id:config.getUUIDFuc(model)}))
            } else {
                dispatch(uploadItemFailed({upload: name}))
            }
            return uploadNextItem(name, dispatch, getState, config)
        })
        .catch((error) => {
            dispatch(uploadItemFailed({upload: name}))
            return uploadNextItem(name, dispatch, getState, config)
        })
}

function getNextUploadItem(name, getState) {
    const state = getState()
    const upload = state.upload.get(name)
    if(!upload || upload.cancle) return null

    return upload.wattingQueue.first()
}

function uploadNextItem(name, dispatch, getState, config) {
    let nextItem = getNextUploadItem(name, getState)
    if(nextItem) {
        return uploadItem(name, nextItem, dispatch, getState, config)
    } else {
        dispatch(uploadComplete({upload: name}))
        //发送事件
        UploadEventEmitter.emitComplete(name)
        return Promise.resolve()
    }
}