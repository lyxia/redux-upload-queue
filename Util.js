//@flow

export function getUploadData(state, upload) {
    const uploadObj = state.upload.get(upload)
    return uploadObj
}

export function getUploadedCount(state, upload) {
    const uploadObj = getUploadData(state, upload)
    if(uploadObj) {
        return uploadObj.uploadedQueue.size
    }
    return 0
}

export function getWattingCount(state, upload) {
    const uploadObj = getUploadData(state, upload)
    if(uploadObj) {
        return uploadObj.wattingQueue.size
    }
    return 0
}

export function getFailedCount(state, upload) {
    const uploadObj = getUploadData(state, upload)
    if(uploadObj) {
        return uploadObj.failedQueue.size
    }
    return 0
}

export function hasAllsccess(state, upload) {
    if(getWattingCount(state, upload) == 0 && getFailedCount(state, upload) == 0) {
        return true
    }
    return false
}

export function getUpdateItemID(uploadObj, name) {
    let id = ''
    if(uploadObj) {
        const uploadedQueue = uploadObj.get('uploadedQueue')
        uploadedQueue.forEach((value) => {
            if(value.get('name') === name) {
                id = value.get('id')
                return 
            }
        })
    }
    return id
}

export function getUpdatedIDs(uploadObj) {
    let ids = []
    if(uploadObj) {
        const uploadedQueue = uploadObj.get('uploadedQueue')
        uploadedQueue.forEach((value) => {
            ids.push(value.get('id'))
        })
    }
    return ids
}