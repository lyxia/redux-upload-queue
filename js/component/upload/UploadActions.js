//@flow

import {createAction} from 'redux-actions'

import {uploadImageWithPost} from '../../networking/FetchData'

import UploadEventEmitter from './UploadEventEmitter'

export const pushUploadItem = createAction('Upload/pushItem')
export const deleteUploadItem = createAction('Upload/deleteItem')
export const startuploadItem = createAction('Upload/startUploadItem')
export const uploadItemFailed = createAction('Upload/failedUploadItem')
export const uploadItemSuccess = createAction('Upload/successUploadItem')

export const startUpload = createAction('Upload/startUpload')
export const uploadComplete = createAction('Upload/uploadComplete')

export const uploadCancle = createAction('Upload/uploadCancle')
export const registerUpload = createAction('Upload/registerUpload')
export const destroyUpload = createAction('Upload/destroyUpload')

export function upload(name) {
    return (dispatch, getState) => {
        const state = getState()
        if(state.upload.get(name)) {
            if(!state.upload.get(name).uploading) {
                //开始上传
                dispatch(startUpload({upload: name}))
                uploadNextItem(name, dispatch, getState)
            }
        }
    }
}

function uploadItem(name, item, dispatch, getState) {
    dispatch(startuploadItem({upload: name}))
    uploadImageWithPost(item.filepath, item.name)
        .then((model) => {
            if(model.error == -1) {
                dispatch(uploadItemSuccess({upload: name, id:model.fileName}))
            } else {
                dispatch(uploadItemFailed({upload: name}))
            }
            uploadNextItem(name, dispatch, getState)
        })
        .catch((error) => {
            dispatch(uploadItemFailed({upload: name}))
            uploadNextItem(name, dispatch, getState)
        })
}

function getNextUploadItem(name, getState) {
    const state = getState()
    const upload = state.upload.get(name)
    if(!upload || upload.cancle) return null

    return upload.wattingQueue.first()
}

function uploadNextItem(name, dispatch, getState) {
    let nextItem = getNextUploadItem(name, getState)
    if(nextItem) {
        uploadItem(name, nextItem, dispatch, getState)
    } else {
        dispatch(uploadComplete({upload: name}))
        //发送事件
        UploadEventEmitter.emitComplete(name)
    }
}