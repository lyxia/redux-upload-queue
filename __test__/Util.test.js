import config, { store } from './UploadConfig'
import * as UploadActions from '../UploadActions'
import * as Util from '../Util'

const uploadName = 'uploadKey'
let state = null
beforeAll(() => {
    fetch.mockResponses(
        [
            JSON.stringify({ error: null, id: '123456' })
        ],
        [
            JSON.stringify({ error: new Error('fail') })
        ],
        [
            JSON.stringify({ error: null, id: '1234567' })
        ],
    )

    store.dispatch(UploadActions.registerUpload({ upload: uploadName }))
    store.dispatch(UploadActions.pushUploadItem({ upload: uploadName, name: 'fileOne', filePath: 'filePathOne' }))
    store.dispatch(UploadActions.pushUploadItem({ upload: uploadName, name: 'fileTwo', filePath: 'filePathTwo' }))
    store.dispatch(UploadActions.pushUploadItem({ upload: uploadName, name: 'fileThree', filePath: 'filePathThree' }))
    return store.dispatch(UploadActions.upload(uploadName, config))
        .then(() => {
            state = store.getState()
            return expect(state).toMatchSnapshot()
        })
})

afterAll(() => {
    store.clearActions()
    fetch.resetMocks()
})

test('upload Util getUploadData test', () => {
    expect(Util.getUploadData(state, uploadName)).toMatchSnapshot()
})

test('upload Util getUploadedCount test', () => {
    expect(Util.getUploadedCount(state, uploadName)).toMatchSnapshot()
})

test('upload Util getWattingCount test', () => {
    expect(Util.getWattingCount(state, uploadName)).toMatchSnapshot()
})

test('upload Util getFailedCount test', () => {
    expect(Util.getFailedCount(state, uploadName)).toMatchSnapshot()
})

test('upload Util hasAllsccess test', () => {
    expect(Util.hasAllsccess(state, uploadName)).toMatchSnapshot()
})

test('upload Util getUpdateItemID test', () => {
    const uploadObj = Util.getUploadData(state, uploadName)
    expect(Util.getUpdateItemID(uploadObj, 'fileOne')).toMatchSnapshot()
})

test('upload Util getUpdatedIDs test', () => {
    const uploadObj = Util.getUploadData(state, uploadName)
    expect(Util.getUpdatedIDs(uploadObj)).toMatchSnapshot()
})