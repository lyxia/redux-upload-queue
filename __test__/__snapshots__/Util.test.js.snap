// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Object {
  "upload": Immutable.Map {
    "uploadKey": Immutable.Upload {
      "wattingQueue": Immutable.List [],
      "uploadedQueue": Immutable.List [
        Immutable.UploadItem {
          "filepath": "filePathOne",
          "state": "uploaded",
          "id": "123456",
          "name": "fileOne",
        },
        Immutable.UploadItem {
          "filepath": "filePathThree",
          "state": "uploaded",
          "id": "1234567",
          "name": "fileThree",
        },
      ],
      "failedQueue": Immutable.List [
        Immutable.UploadItem {
          "filepath": "filePathTwo",
          "state": "failed",
          "id": null,
          "name": "fileTwo",
        },
      ],
      "uploading": false,
      "cancle": false,
      "name": "uploadKey",
    },
  },
}
`;

exports[`upload Util getFailedCount test 1`] = `1`;

exports[`upload Util getUpdateItemID test 1`] = `"123456"`;

exports[`upload Util getUpdatedIDs test 1`] = `
Array [
  "123456",
  "1234567",
]
`;

exports[`upload Util getUploadData test 1`] = `
Immutable.Upload {
  "wattingQueue": Immutable.List [],
  "uploadedQueue": Immutable.List [
    Immutable.UploadItem {
      "filepath": "filePathOne",
      "state": "uploaded",
      "id": "123456",
      "name": "fileOne",
    },
    Immutable.UploadItem {
      "filepath": "filePathThree",
      "state": "uploaded",
      "id": "1234567",
      "name": "fileThree",
    },
  ],
  "failedQueue": Immutable.List [
    Immutable.UploadItem {
      "filepath": "filePathTwo",
      "state": "failed",
      "id": null,
      "name": "fileTwo",
    },
  ],
  "uploading": false,
  "cancle": false,
  "name": "uploadKey",
}
`;

exports[`upload Util getUploadedCount test 1`] = `2`;

exports[`upload Util getWattingCount test 1`] = `0`;

exports[`upload Util hasAllsccess test 1`] = `false`;
